ntp-proto/src/nts_record.rs

```diff
@@ -576,30 +640,53 @@ impl KeyExchangeClient {
-                            ) {
-                                return ControlFlow::Break(Err(KeyExchangeError::Tls(e)));
-                            }
-                            let mut s2c: aead::Key<Aes128SivAead> = Default::default();
-                            if let Err(e) = self.tls_connection.export_keying_material(
-                                &mut s2c,
-                                Self::NTS_KEY_EXPORT_LABEL,
-                                Some(Self::NTS_KEY_EXPORT_AESSIV128_S2C_CONTEXT),
-                            ) {
-                                return ControlFlow::Break(Err(KeyExchangeError::Tls(e)));
-                            }
+                            let algorithm = result.algorithm.unwrap_or_default();
+
+                            tracing::info!(?algorithm, "selected AEAD algorithm");
+
+                            let nts = match algorithm {
+                                AeadAlgorithm::AeadAesSivCmac256 => {
+                                    let keys = extract_nts_keys::<Aes128SivAead>(
+                                        &self.tls_connection,
+                                        algorithm,
+                                    );
+
+                                    let NtsKeys { c2s, s2c } =
+                                        match keys.map_err(KeyExchangeError::Tls) {
+                                            Ok(keys) => keys,
+                                            Err(e) => return ControlFlow::Break(Err(e)),
+                                        };
+
+                                    PeerNtsData {
+                                        cookies: result.cookies,
+                                        c2s: Box::new(AesSivCmac256::new(c2s)),
+                                        s2c: Box::new(AesSivCmac256::new(s2c)),
+                                    }
+                                }
+                                AeadAlgorithm::AeadAesSivCmac512 => {
+                                    let keys = extract_nts_keys::<Aes256SivAead>(
+                                        &self.tls_connection,
+                                        algorithm,
+                                    );
+
+                                    let NtsKeys { c2s, s2c } =
+                                        match keys.map_err(KeyExchangeError::Tls) {
+                                            Ok(keys) => keys,
+                                            Err(e) => return ControlFlow::Break(Err(e)),
+                                        };
+
+                                    PeerNtsData {
+                                        cookies: result.cookies,
+                                        c2s: Box::new(AesSivCmac512::new(c2s)),
+                                        s2c: Box::new(AesSivCmac512::new(s2c)),
+                                    }
+                                }
+                            };
```

Comment:
This is getting rather bulky, I think we should at least refactor this into it's own function, and perhaps try to find a way to refactor more of the repetition out of this.

Maybe pushing this onto AeadAlgorithm is a good idea?

Reply:
yes, refactored this and it looks nicer now